### PR TITLE
LOOP-1114: adds Version Checking to the responsibilities of `SupportUI` [plugins]

### PR DESCRIPTION
**Pull request by rickpasetto**
https://tidepool.atlassian.net/browse/LOOP-1114

After a few rounds of implementation, we settled on having the existing `SupportUI` plugin interface support the ability to do version checking.  So, now it has the responsibilities to:
- Do a version check with whatever "service" it needs to to determine if a new version of the host app is available
- Provide a UI screen to explain to the user what kind of update it is (from **Settings**)
- Provide a menu item (like it did before) for Support

This PR also pulls a bunch of functionality out of `MockService` and into a new `MockSupport` class for testing.

LW PR: https://github.com/tidepool-org/LoopWorkspace/pull/670

### CODE REVIEW

File: LoopKitUI/SupportUI.swift [hunk 1]
```diff
@@ -28,4 +30,45 @@ public protocol SupportUI {
...
+    /// Check whether the given app version for the given `bundleIdentifier` needs an update.  Services should return their last result, if known.
+    ///
+    /// - Parameters:
+    ///    - bundleIdentifier: The host app's `bundleIdentifier` (a.k.a. `CFBundleIdentifier`) string.
+    ///    - currentVersion: The host app's current version (i.e. `CFBundleVersion`).
+    ///    - completion: The completion function to call with any success result (or `nil` if not known) or failure.
+    func checkVersion(bundleIdentifier: String, currentVersion: String, completion: @escaping (Result<VersionUpdate?, Error>) -> Void)
+
+    /// Hand this support an `AlertIssuer` so it may be able to issue alerts.
+    func setAlertIssuer(alertIssuer: AlertIssuer?)
```
Comment (darinkrauss) [inline LoopKitUI/SupportUI.swift]: This seems non-standard. The pump and CGM managers use a delegate as the alert issuer plus they are `AlertResponder`s.

File: LoopKitUI/SupportUI.swift [hunk 2]
```diff
@@ -28,4 +30,45 @@ public protocol SupportUI {
...
+    
+    /// Provides screen for software update UI.
+    ///
+    /// - Parameters:
+    ///    - guidanceColors: Colors to use for warnings, etc.
+    ///    - bundleIdentifier: The host app's bundle identifier (e.g. `Bundle.main.bundleIdentifier`).
+    ///    - currentVersion: The host app's current version (i.e. `CFBundleVersion`).
+    ///    - openAppStoreHook: Hook function to open up the App Store for the host app.
+    /// - Returns: A view that will be opened when a software update is available from this service.
+    func softwareUpdateView(guidanceColors: GuidanceColors,
```
Comment (darinkrauss) [inline LoopKitUI/SupportUI.swift]: [nit] I would put `guidanceColors` last as it is the least relevant to this particular view.

File: LoopKitUI/SupportUI.swift [hunk 3]
```diff
@@ -28,4 +30,45 @@ public protocol SupportUI {
...
+                            bundleIdentifier: String,
+                            currentVersion: String,
+                            openAppStoreHook: (() -> Void)?
```
Comment (darinkrauss) [inline LoopKitUI/SupportUI.swift]: [nit] "Hook"?

File: LoopKitUI/SupportUI.swift [hunk 4]
```diff
@@ -28,4 +30,45 @@ public protocol SupportUI {
...
+    ) -> AnyView?
```
Comment (darinkrauss) [inline LoopKitUI/SupportUI.swift:57]: Do we really want to be passing SwiftUI through these protocols? We don't do that anywhere else, AFAIK.

File: MockKit/MockCGMManager.swift [hunk 5]
```diff
@@ -329,8 +329,13 @@ public final class MockCGMManager: TestingCGMManager {
                                              foregroundContent: Alert.Content(title: "Signal Loss", body: "CGM simulator signal loss", acknowledgeActionButtonLabel: "Dismiss"),
                                              backgroundContent: Alert.Content(title: "Signal Loss", body: "CGM simulator signal loss", acknowledgeActionButtonLabel: "Dismiss"))
 
+    private let lockedMockSensorState = Locked(MockCGMState(isStateValid: true))
```
Comment (darinkrauss) [inline MockKit/MockCGMManager.swift]: These changes look like there are for https://tidepool.atlassian.net/browse/LOOP-3884. Consider a different PR or, at the very minimum, reference this ticket in the commits and PR.

File: MockKitUI/View Controllers/MockServiceTableViewController.swift [hunk 6]
```diff
@@ -357,35 +339,23 @@ fileprivate extension UIAlertController {
...
-                  preferredStyle: .actionSheet
-        )
-        for versionUpdate in VersionUpdate.allCases {
-            addAction(UIAlertAction(title: versionUpdate.localizedDescription, style: versionUpdate.actionStyle, handler: { _ in handler(versionUpdate) }))
-        }
-        addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
-    }
 }
 
 fileprivate extension VersionUpdate {
```
Comment (darinkrauss) [inline MockKitUI/View Controllers/MockServiceTableViewController.swift]: This should probably be elsewhere if you just removed all of the other update related code from the `MockServiceTableViewController`.

File: MockKitUI/MockSupport.swift [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        }
+        
+        let alertIdentifier = Alert.Identifier(managerIdentifier: MockSupport.supportIdentifier, alertIdentifier: versionUpdate.rawValue)
+        let alertContent: LoopKit.Alert.Content
+        if firstAlert {
+            alertContent = Alert.Content(title: versionUpdate.localizedDescription,
+                                         body: NSLocalizedString("""
+                                                Your app is out of date. It will continue to work, but we recommend updating to the latest version.
+                                                """, comment: "Alert content body for first software update alert"),
+                                         acknowledgeActionButtonLabel: NSLocalizedString("OK", comment: "default acknowledgement"),
```
Comment (darinkrauss) [inline MockKitUI/MockSupport.swift]: [nit] Capitalize first word of comment.

File: MockKitUI/MockSupport.swift [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+                                         isCritical: versionUpdate == .required)
+        } else if let lastVersionCheckAlertDate = lastVersionCheckAlertDate,
+                  abs(lastVersionCheckAlertDate.timeIntervalSinceNow) > alertCadence {
+            alertContent = Alert.Content(title: NSLocalizedString("Update Reminder", comment: "Recurring software update alert title"),
+                                         body: NSLocalizedString("""
+                                                Your app is still out of date. It will continue to work, but we recommend updating to the latest version.
+                                                """, comment: "Alert content body for recurring software update alert"),
+                                         acknowledgeActionButtonLabel: NSLocalizedString("OK", comment: "default acknowledgement"),
```
Comment (darinkrauss) [inline MockKitUI/MockSupport.swift]: [nit] Capitalize first word of comment.

File: MockKitUI/MockSupport.swift [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+                                         isCritical: versionUpdate == .required)
+        } else {
+            return
+        }
+        alertIssuer!.issueAlert(Alert(identifier: alertIdentifier, foregroundContent: alertContent, backgroundContent: alertContent, trigger: .immediate))
```
Comment (darinkrauss) [inline MockKitUI/MockSupport.swift]: If `alertIssuer` is optional (see `setAlertIssuer(alertIssuer: AlertIssuer?)`), then should we really force it here? Either make `alertIssuer` required or make this `alertIssuer?`.

File: MockKitUI/MockSupport.swift [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    
+    let mockSupport: MockSupport
+    
+    @State var showActionSheet: Bool = false
+    
+    private var buttons: [ActionSheet.Button] {
+        VersionUpdate.allCases.map { versionUpdate in
+            let setter = { mockSupport.versionUpdate = versionUpdate }
+            switch versionUpdate {
+            case .required: return ActionSheet.Button.destructive(Text(versionUpdate.localizedDescription), action: setter)
```
Comment (darinkrauss) [inline MockKitUI/MockSupport.swift]: I'm am not a fan of having the statement on the same line as the case (I find it much easier to miss the statement being executed), but I can deal with it when it is trivial code. This, however, is not trivial, so please move to separate lines. And, really, what exactly are you saving here by putting them on the same line?

**Review by darinkrauss — CHANGES_REQUESTED** [review]
A few changes needed here.

File: LoopKitUI/SupportUI.swift (hunk 1 above)
Comment (rickpasetto) [reply LoopKitUI/SupportUI.swift]: Yeah I am not sure I would call that a standard.  That was adding Alerting to the plugins within the existing plugin framework.  I'll follow that pattern, though, as you suggest. (TBH I prefer `AlertIssuer`.  It is a clearer name & responsibility than just `SupportUIDelegate`)

File: LoopKitUI/SupportUI.swift (hunk 2 above)
Comment (rickpasetto) [reply LoopKitUI/SupportUI.swift]: It's a Swift standard to put function blocks parameters last in functions if possible.  But I will put it before that.

File: LoopKitUI/SupportUI.swift (hunk 3 above)
Comment (rickpasetto) [reply LoopKitUI/SupportUI.swift]: I'll remove

File: LoopKitUI/SupportUI.swift (hunk 4 above)
Comment (rickpasetto) [reply LoopKitUI/SupportUI.swift:57]: Why not?  I see this as an opportunity to move towards SwiftUI... other "ManagerUI" protocols return UIKit ViewControllers.  What's the down side?

File: MockKit/MockCGMManager.swift (hunk 5 above)
Comment (rickpasetto) [reply MockKit/MockCGMManager.swift]: A separate PR has been merged

File: LoopKitUI/SupportUI.swift (hunk 4 above)
Comment (darinkrauss) [reply LoopKitUI/SupportUI.swift:57]: Probably nothing, but you should get @ps2's take on it.

Comment (ps2) [reply LoopKitUI/SupportUI.swift:57]: I don't see a problem with this, per se. The only nit I see is that you have to use AnyView, but this isn't a blocker.

File: LoopKitUI/SupportUI.swift [hunk 11]
```diff
@@ -17,9 +17,15 @@ public protocol SupportInfoProvider {
     func generateIssueReport(completion: @escaping (String) -> Void)
 }
 
-public protocol SupportUI {
+public protocol SupportUIDelegate: AlertIssuer {
```
Comment (darinkrauss) [inline LoopKitUI/SupportUI.swift]: [nit] Collapse into one line?

**Review by darinkrauss — APPROVED** [review]
LGTM!

**Pull request merged**